### ISSUE TITLE
Some drivers with saul capabilities are not tested automatically by the saul test

### ISSUE THREAD

**Issue by aabadie** [RIOT-OS/RIOT#6405]
This is the case for:
* si70xx
* bmp180
* jc42
* and soon bme280 (after #6156 is merged)

I open this issue to discuss a solution that could be applied to the CI scripts : [dist/tools/compile_test/compile_test.py](https://github.com/RIOT-OS/RIOT/blob/master/dist/tools/compile_test/compile_test.py)) or the [JenkinsFile](https://github.com/RIOT-OS/RIOT/blob/master/Jenkinsfile).

@PeterKietzmann suggested a simple solution in this [comment](https://github.com/RIOT-OS/RIOT/pull/6156#issuecomment-272965500), e.g call building `tests/saul` for each sensor drivers having a saul implementation, to verify it builds. For example:
```bash
drivers="bmp180 jc42 si70xx"
for driver in $drivers; do USEMODULE=$driver make -C tests/saul clean all; done
```

**Comment by aabadie** [comment]
@smlng, some of your PRs yesterday made me remember this issue. In `made buidtests`, the idea would be to check if the application to be built is `examples/saul` and loop on all drivers in `sys/auto_init/saul`. Then the SAUL build of drivers will be part of the CI.

What do you think of the idea ?

I'd like to look at it for this release.

**Comment by smlng** [comment]
@aabadie had the same idea yesterday, when working on those SAUL refinements - and basically I did as you suggested: take `examples/saul` and compile every SAUL capable driver like:

```
USEMODULE=<driver-module> BOARD=samr21-xpro make -C examples/saul clean all
```

The question is how to proceed and make it future prove? Integrate with `make buildtests` wouldn't be _my_ favorite, nevertheless, better than nothing.

**Comment by kaspar030** [comment]
> USEMODULE=<driver-module> BOARD=samr21-xpro make -C examples/saul clean all

There are test applications for most drivers. How about just including saul in them?

While the CI could do "for driver in drivers; do USEMODULE=driver make -C examples/saul clean all", it would completely break both ccache and the git-caching scheme used to avoid a new checkout for each build.

**Comment by smlng** [comment]
> There are test applications for most drivers. How about just including saul in them?

Though that might be the easiest way, I don't think its a good idea. Because most driver tests are functional tests, and would then break when somethings wrong with SAUL - or the other way around.

Further, I'm huge _fan_ of keeping tests small and simple - so they (hopefully) fit and execute on even the smallest platforms, like Arduino Uno (remember: just 2K of RAM).

**Comment by aabadie** [comment]
Maybe we could add some logic for getting saul capable drivers in compile_test.py ?

This won't solve the ccache issue if I understand correctly.

**Comment by smlng** [comment]
For #7241 (by @kaspar030 ) to work, each drivers has to state its required features. Could we extend (or reverse) that logic so that drivers also can state which features they provide, i.e., SAUL for instance? 

And if that's possible, could we exploit that for the tests and/or CI building process to check SAUL if a driver implements SAUL interface?

**Comment by aabadie** [comment]
In python, this can be done in one line (PEP8 inside):
```python
drivers = [os.path.splitext(i)[0].split('_')[-1] for i in os.listdir('sys/auto_init/saul') if i != 'Makefile' and i not in  ('auto_init_adc.c', 'auto_init_gpio.c')]
```

**Comment by kaspar030** [comment]
How about a "saul_drivers" test application, building *all* saul enabled drivers? (This is about compiling, right?)

**Comment by aabadie** [comment]
> How about a "saul_drivers" test application, building all saul enabled drivers?

The problem will be with low memory boards.

**Comment by smlng** [comment]
Yes, a compile test would be enough for the beginning. Still would be nice to have an automatic way to get _all_ SAUL enabled drivers, and not adapt the test for every new SAUL adaption.

**Comment by miri64** [comment]
> The problem will be with low memory boards.

They still compile when they are in `BOARD_INSUFFICIENT_MEMORY`, just the linking is skipped.

**Comment by aabadie** [comment]
Ok, then let's go for the saul_drivers application. I'll try to push a PR later today.

**Comment by aabadie** [comment]
This issue is still valid. Maybe the underlying idea of #11676 could help a bit here.
